socket_wrapper.hpp

```diff
@@ -2,6 +2,11 @@
 #include "util.h"
 #include <memory>
 
+class SocketWrapper;
+
+using SocketWrapper_uptr = std::unique_ptr<SocketWrapper>;
+using SocketWrapper_sptr = std::shared_ptr<SocketWrapper>;
+
 class SocketWrapper
```

Comment:
Just as a niggle: I'd better call this class simply `Socket`. As it will duplicate corresponding functionality in OOP way, I think it's more-or-less obvious that it is just a wrapper.
But it is fully up to you, just let me know if you'll prefer to left current name.